### PR TITLE
CAT-1276 Introduce pmd controversial ruleset

### PR DESCRIPTION
**Pull request by cheidenreich**
Jenkins run:
https://jenkins.catrob.at/view/All-Categories/view/Catroid-multi-job/job/Catroid-Multi-Job-Custom-Branch-RELOADED/2204/

UserBrickScriptActivityTest single run:
https://jenkins.catrob.at/view/All-Categories/view/Catroid/job/Catroid-CustomEmulator-Single-ClassAndPackage-Test/445/

Checkstyle waring has already been fixed through force push

### CODE REVIEW

**Comment by cheidenreich** [comment]
the testAddNewLook seems to fail on current develop since yesterday: https://jenkins.catrob.at/view/All-Categories/view/Catroid/job/Catroid-Device-Single-ClassAndPackage-Test/355/

File: catroid/src/org/catrobat/catroid/bluetooth/ConnectBluetoothDeviceActivity.java [hunk 1]
```diff
@@ -152,7 +152,7 @@ public void onReceive(Context context, Intent intent) {
 
 	private class ConnectDeviceTask extends AsyncTask<String, Void, BluetoothConnection.State> {
 
-		BluetoothConnection btConnection;
+		private BluetoothConnection btConnection;
```
Comment (robertpainsi) [inline catroid/src/org/catrobat/catroid/bluetooth/ConnectBluetoothDeviceActivity.java]: https://github.com/Catrobat/Catroid/pull/1216#discussion_r30360929
The check has to be excluded.

File: catroid/src/org/catrobat/catroid/common/LookData.java [hunk 2]
```diff
@@ -114,9 +114,6 @@ public Pixmap getOriginalPixmap() {
 		return originalPixmap;
 	}
 
-	public LookData() {
-	}
```
Comment (robertpainsi) [inline catroid/src/org/catrobat/catroid/common/LookData.java]: https://github.com/Catrobat/Catroid/pull/1216#issuecomment-121588850
The check has to be excluded.

**Comment by robertpainsi** [comment]
Take a look at `git reset --soft` and `git add -p` if you aren't familiar with them yet. You should be able to clean up the PR without opening any editor :wink:

Since adding private (not public!) to members and methods without access modifiers is a good idea, create another PR for that.

**Comment by cheidenreich** [comment]
testAddNewLook is still failing on develop

**Comment by robertpainsi** [comment]
Looks good to me!
I'll just take a quick look over the whole controversial ruleset tomorrow to check if I didn't miss anything. Otherwise it's ready to merge ^^

File: config/pmd_controversial.xml [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        <exclude name="UseConcurrentHashMap" />
+        <exclude name="UseObjectForClearerAPI" />
+        <exclude name="AvoidUsingShortType" />
+        <exclude name="DefaultPackage" />
+        <exclude name="UnnecessaryConstructor" />
+        <exclude name="DoNotCallGarbageCollectionExplicitly">
+            <exclude-pattern>.*src\org\catrobat\catroid\ui\BaseActivity.java</exclude-pattern>
+        </exclude>
+        <exclude name="AvoidUsingVolatile">
+            <exclude-pattern>.*src\org\catrobat\catroid\content\Script.java</exclude-pattern>
```
Comment (robertpainsi) [inline config/pmd_controversial.xml]: [This](https://github.com/Catrobat/Catroid/blob/e0af7413aee3f28f4cb1cd210b48ea998934ec71/catroid/src/org/catrobat/catroid/content/Script.java#L51) is the only volatile variable which in the meantime is a relic of the old thread-based handling of actions.
The volatile modifier has its use-cases and this rule should be excluded completely.
BUT I'm not familiar with the Android asynch tasks/jobs/etc but they may also rely on the volatile modifier, I guess?

**Comment by robertpainsi** [comment]
ping @avihric @obusher @thmq 
May someone of you also take a look at the new pmd rules please? Since a pmd warning will abort the whole jenkins testrun now, people will more likely try to make jenkins happy and decrease code quality instead of adapting the rule sets
https://pmd.github.io/pmd-5.3.3/pmd-java/rules/java/controversial.html

**Comment by avihric** [comment]
> ... decrease code quality

I guess you mean `increase` :stuck_out_tongue:

**Comment by robertpainsi** [comment]
Au contraire! :joy_cat:

What I wanted to say in a very (too ^^) short sentence, was a scenario like this:
Checks like [OnlyOneReturn](https://pmd.github.io/pmd-5.3.3/pmd-java/rules/java/controversial.html#OnlyOneReturn) or [CallSuperInConstructor](https://pmd.github.io/pmd-5.3.3/pmd-java/rules/java/controversial.html#CallSuperInConstructor) unnecessarily bloats the code and decreases code quality. So if we add such checks unintentionally, people still will only try to make jenkins happy by using just a single return or always call super in constructor instead of excluding these rules from pmd.

**Comment by robertpainsi** [comment]
I will merge it asap. Just fix the last complain and rerun jenkins :sunglasses:
A quick rebase would be awesome as well. :wink:

**Comment by cheidenreich** [comment]
done

**Comment by robertpainsi** [comment]
Thanks! :smile_cat:

**Pull request merged**